source/fuzz/transformation_propagate_instruction_down.h

```diff
@@ -0,0 +1,136 @@
+  //   unique and fresh.
+  bool IsApplicable(
+      opt::IRContext* ir_context,
+      const TransformationContext& transformation_context) const override;
+
+  // - Adds a clone of the propagated instruction into every acceptable
+  //   successor of |block_id|.
+  // - Removes the original instruction.
+  // - Creates an OpPhi instruction if possible, that tries to group created
+  //   clones.
+  // - If the original instruction's id was irrelevant - marks created
+  //   instructions as irrelevant. Otherwise, marks the created instructions as
+  //   synonymous to each other.
+  void Apply(opt::IRContext* ir_context,
+             TransformationContext* transformation_context) const override;
+
+  protobufs::Transformation ToMessage() const override;
+
+  // Returns true if this transformation can be applied to the block with id
+  // |block_id|. Concretely, returns true iff:
+  // - |block_id| is a result id of some reachable basic block in the module.
+  // - the block has an instruction to propagate (see
+  //   GetInstructionToPropagate method).
+  // - the block has at least one acceptable successor (see
+  //   GetAcceptableSuccessors method).
+  // - none of the acceptable successors have OpPhi instructions that use the
+  //   original instruction.
+  // - it is possible to replace every use of the original instruction with some
+  //   of the propagated instructions (or an OpPhi if we can create it - see
+  //   CanAddOpPhiInstruction method).
+  static bool IsApplicableToBlock(opt::IRContext* ir_context,
+                                  uint32_t block_id);
+
+  // Returns ids of successors of |block_id|, that can be used to propagate an
+  // instruction into. Concretely, a successor block is acceptable if all
+  // dependencies of the propagated instruction dominate it. Note that this
+  // implies that an acceptable successor must be reachable in the CFG.
+  // For example:
+  //    %1 = OpLabel
+  //         OpSelectionMerge %2 None
+  //         OpBranchConditional %cond %2 %3
+  //    %3 = OpLabel
+  //    %4 = OpUndef %int
+  //    %5 = OpCopyObject %int %4
+  //         OpBranch %2
+  //    %2 = OpLabel
+  //    ...
+  // In this example, %2 is not an acceptable successor of %3 since one of the
+  // dependencies (%4) of the propagated instruction (%5) does not dominate it.
+  static std::unordered_set<uint32_t> GetAcceptableSuccessors(
+      opt::IRContext* ir_context, uint32_t block_id);
+
+  std::unordered_set<uint32_t> GetFreshIds() const override;
+
+ private:
+  // Returns the last possible instruction in the |block_id| that satisfies the
+  // following properties:
+  // - has result id
+  // - has type id
+  // - has supported opcode (see IsOpcodeSupported method)
+  // - has no users in its basic block.
+  // Returns nullptr if no such an instruction exists.
+  static opt::Instruction* GetInstructionToPropagate(opt::IRContext* ir_context,
+                                                     uint32_t block_id);
```

Comment:
Would you be able to give a small example here of a basic block with some instructions, such that the last 2-3 are not suitable, and explain why?

Reply:
Fixed in b6fe77b.